### PR TITLE
Importer option for permanent scaling of the scene

### PR DESCRIPTION
**Pull request by SaracenOne**
This is a draft PR tied to a proposal to replace or provide an alternative to the existing importer root_scale option (https://github.com/godotengine/godot-proposals/issues/4379). The root_scale option only provides a one off modification to the root node's scale value when its instantiated, whereas this option provides a permentant modification to the underlying geometry and animations of a scene.

Left in draft while proposal is considered and the PR is being throughly tested for bugs.

Thanks to @lyuma for the original script implementation of this which I used as reference when writing this C++ version.

### CODE REVIEW

File: editor/import/resource_importer_scene.cpp [hunk 1]
```diff
@@ -372,6 +372,194 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+
+	for (int surf_idx = 0; surf_idx < surf_count; surf_idx++) {
+		Mesh::PrimitiveType prim = surf_data_by_mesh[surf_idx].prim;
+		Array arr = surf_data_by_mesh[surf_idx].arr;
+		Array bsarr = surf_data_by_mesh[surf_idx].bsarr;
+		Dictionary lods = surf_data_by_mesh[surf_idx].lods;
+		int fmt_compress_flags = surf_data_by_mesh[surf_idx].fmt_compress_flags;
+		String name = surf_data_by_mesh[surf_idx].name;
+		Ref<Material> mat = surf_data_by_mesh[surf_idx].mat;
+		// print("Adding mesh vertices by " + str(p_scale) + ": " + str(arr[ArrayMesh.ARRAY_VERTEX][0]))
```
Comment (fire) [inline editor/import/resource_importer_scene.cpp:458]: Comments should be removed or made `print_verbose`.

**Comment by fire** [comment]
Typo on the commit message. `permentant` -> `permanent`

File: editor/import/resource_importer_scene.cpp [hunk 2]
```diff
@@ -1537,6 +1725,7 @@ void ResourceImporterScene::get_import_options(const String &p_path, List<Import
 		script_ext_hint += "*." + E;
 	}
 
+	r_options->push_back(ImportOption(PropertyInfo(Variant::FLOAT, "nodes/scene_scale", PROPERTY_HINT_RANGE, "0.001,1000,0.001"), 1.0));
```
Comment (marstaik) [inline editor/import/resource_importer_scene.cpp:1728]: Why introduce a new scale option? Why not a toggle for "Bake root scale" ?

**Review by marstaik — COMMENTED** [review]
I see a lot of unnecessary data copies, and a lack of const correctness. Can you go through and check on them? A lot of the data copies will not be able to be solved due to lack of move semantics however.

File: editor/import/resource_importer_scene.cpp [hunk 3]
```diff
@@ -372,6 +372,194 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+	}
+
+	if (!is_shadow && p_mesh->get_shadow_mesh() != p_mesh && p_mesh->get_shadow_mesh().is_valid()) {
+		_rescale_importer_mesh(p_scale, p_mesh->get_shadow_mesh(), true);
+	}
+}
+
+void _rescale_skin(float p_scale, Ref<Skin> p_skin) {
+	// MESH and SKIN data divide, to compensate for object position multiplying.
+	for (int i = 0; i < p_skin->get_bind_count(); i++) {
```
[the remark this answers is not in the document]
Comment (lyuma) [reply editor/import/resource_importer_scene.cpp:470]: I actually disagree that this is dictated by Godot's style guide `i++` is far more common than `++i` in the Godot codebase.
```
$ cgrep -r '++[a-zA-Z_][a-zA-Z_0-9]*).*{' . --exclude-dir=thirdparty |wc -l
548

$ cgrep -r '[a-zA-Z_][a-zA-Z_0-9]*++).*{' . --exclude-dir=thirdparty |wc -l
8614
```

File: editor/import/resource_importer_scene.cpp [hunk 4]
```diff
@@ -372,6 +372,194 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
 	}
 }
 
+struct ScalableNodeCollection {
+	List<Node3D *> node_3ds;
```
[the remark this answers is not in the document]
Comment (lyuma) [reply editor/import/resource_importer_scene.cpp:376]: Good catch that heap List is not appropriate.

But I actually believe these should be Set<> because it is doing .find() followed by .push_back() which will cause an O(N^2) order complexity.

Set is fine because processing data can happen in arbitrary order

File: editor/import/resource_importer_scene.cpp [hunk 5]
```diff
@@ -372,6 +372,194 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+	List<Ref<Animation>> animations;
+};
+
+void _rescale_importer_mesh(float p_scale, Ref<ImporterMesh> p_mesh, bool is_shadow = false) {
+	// MESH and SKIN data divide, to compensate for object position multiplying.
+
+	int surf_count = p_mesh->get_surface_count();
+	struct LocalSurfData {
+		Mesh::PrimitiveType prim;
+		Array arr;
```
[the remark this answers is not in the document]
Comment (lyuma) [reply editor/import/resource_importer_scene.cpp:388]: The naming `arr` comes from:
```
Array arr = p_mesh->get_surface_arrays(surf_idx);
```
This is currently the conventional name in Godot to use for Array objects holding the surface vertex and index buffers. Other naming suggestions may be welcome but this would affect a lot of import or mesh code.

File: editor/import/resource_importer_scene.cpp (hunk 2 above)
Comment (lyuma) [reply editor/import/resource_importer_scene.cpp:1728]: This sort of discussion might be better suited for the proposal itself. I think the idea was to remove or deprecate the old scene scale which would make the checkbox redundant.

File: editor/import/resource_importer_scene.cpp [hunk 6]
```diff
@@ -372,6 +372,194 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+			Array current_bsarr = p_mesh->get_surface_blend_shape_arrays(surf_idx, bsidx);
+			Vector<Vector3> current_bs_vertex_array = current_bsarr[ArrayMesh::ARRAY_VERTEX];
+
+			int current_bs_vert_arr_len = current_bs_vertex_array.size();
+			i = 0;
+			while (i < current_bs_vert_arr_len) {
+				current_bs_vertex_array.set(i, current_bs_vertex_array[i] * p_scale);
+				i += 1;
+			}
+			current_bsarr[ArrayMesh::ARRAY_VERTEX] = current_bs_vertex_array;
```
Comment (lyuma) [inline editor/import/resource_importer_scene.cpp:431]: This code is really messy.

I did things this way in my GDScript implementation because we do not currently have access to a better API for modifying ImporterMesh.

But since we're making a core engine change, it might be better to provide access to the vertex positions of EditorSceneImporterMesh so that they can be modified without deleting and readding everything.

Here's or chance to improve the APIs to make them less clunky.

**Review by lyuma — COMMENTED** [review]
Overall looking ok.

I do wonder... do we need to consider other objects types for rescaling? For example, light attenuation? camera scale/near/far?

**Comment by fire** [comment]
Want to rebase and try resolving the review comments?

**Comment by fire** [comment]
According to @lyuma this resolves the blend shape bug. Will test.

File: editor/import/resource_importer_scene.cpp (hunk 2 above)
Comment (fire) [reply editor/import/resource_importer_scene.cpp:1727]: Resolved because it should be in a separate pr.

**Comment by TokageItLab** [comment]
Can this function be applied to rotations as well? I was thinking of something similar in #63662. I prefer that those transformations need to be PreProcess for retargeting.

**Comment by fire** [comment]
I want to propose to set the root transform as the default value of this importer option.

**Comment by SaracenOne** [comment]
I'm opening this up for review again based on the changes implemented by @lyuma and @fire

File: editor/import/resource_importer_scene.cpp [hunk 7]
```diff
@@ -1678,6 +1871,7 @@ void ResourceImporterScene::get_import_options(const String &p_path, List<Import
 		script_ext_hint += "*." + E;
 	}
 
+	r_options->push_back(ImportOption(PropertyInfo(Variant::BOOL, "nodes/apply_scene_root"), true));
```
Comment (reduz) [inline editor/import/resource_importer_scene.cpp:1874]: I would rename to 'apply_root_scale'

**Review by reduz — CHANGES_REQUESTED** [review]

File: editor/import/resource_importer_scene.cpp (hunk 7 above)
Comment (fire) [reply editor/import/resource_importer_scene.cpp]: I renamed to 'apply_root_basis'. Hope that fits.

File: editor/import/resource_importer_scene.cpp [hunk 8]
```diff
@@ -368,6 +368,196 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+		if (node_3d) {
+			node_3d->set_position(p_basis.xform(node_3d->get_position()));
+
+			Skeleton3D *skeleton_3d = Object::cast_to<Skeleton3D>(node_3d);
+			if (skeleton_3d) {
+				for (int i = 0; i < skeleton_3d->get_bone_count(); i++) {
+					Transform3D rest = skeleton_3d->get_bone_rest(i);
+					skeleton_3d->set_bone_rest(i, Transform3D(rest.basis, p_basis.xform(rest.origin)));
+					skeleton_3d->set_bone_pose_position(i, p_basis.xform(rest.origin));
+				}
```
Comment (TokageItLab) [inline editor/import/resource_importer_scene.cpp:503]:
```suggestion
				Vector<int> bones_to_process = skeleton_3d->get_parentless_bones();
				for (int i = 0; i < bones_to_process.size(); i++) {
					skeleton_3d->set_bone_rest(bones_to_process[i], p_basis.orthonormalized() * src_skeleton->get_bone_rest(bones_to_process[i]));
				}
				Vector3 scl = p_basis.get_scale_local();
				while (bones_to_process.size() > 0) {
					int idx = bones_to_process[0];
					bones_to_process.erase(idx);
					Vector<int> children = skeleton_3d->get_bone_children(idx);
					for (int i = 0; i < children.size(); i++) {
						bones_to_process.push_back(children[i]);
					}
					skeleton_3d->set_bone_rest(idx, Transform3D(skeleton_3d->get_bone_rest(idx).basis, skeleton_3d->get_bone_rest(idx).origin * scl));
				}
```

File: editor/import/resource_importer_scene.cpp [hunk 9]
```diff
@@ -368,6 +368,196 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+	// MESH and SKIN data divide, to compensate for object position multiplying.
+	for (int i = 0; i < p_skin->get_bind_count(); i++) {
+		Transform3D transform = p_skin->get_bind_pose(i);
+		p_skin->set_bind_pose(i, Transform3D(transform.basis, p_basis.xform(transform.origin)));
+	}
+}
+
+void _rescale_animation(Basis p_basis, Ref<Animation> p_animation) {
+	for (int track_idx = 0; track_idx < p_animation->get_track_count(); track_idx++) {
+		if (p_animation->track_get_type(track_idx) == Animation::TYPE_POSITION_3D) {
```
Comment (TokageItLab) [inline editor/import/resource_importer_scene.cpp:482]: As indicated above, the Rotation3DTrack must also be modified.

**Comment by fire** [comment]
I'll see if I can reset the branch to avoid the big diffs.

File: editor/import/resource_importer_scene.cpp (hunk 9 above)
Comment (TokageItLab) [reply editor/import/resource_importer_scene.cpp:482]: Agreed with @lyuma and @fire that rotation should be done in #63662, not here.

File: modules/gridmap/grid_map.cpp [hunk 10]
```diff
@@ -572,11 +572,12 @@ bool GridMap::_octant_update(const OctantKey &p_key) {
 				NavigationServer3D::get_singleton()->region_set_map(region, get_world_3d()->get_navigation_map());
 				nm.region = region;
 
+#ifdef DEBUG_ENABLED
```
Comment (akien-mga) [inline modules/gridmap/grid_map.cpp:575]: Is this related to this PR?

Comment (fire) [reply modules/gridmap/grid_map.cpp:575]: This is related a clang error. I can remove since it only shows on my os.

File: editor/import/resource_importer_scene.cpp [hunk 11]
```diff
@@ -368,6 +368,196 @@ static void _pre_gen_shape_list(Ref<ImporterMesh> &mesh, Vector<Ref<Shape3D>> &r
...
+		Array arr = p_mesh->get_surface_arrays(surf_idx);
+		String name = p_mesh->get_surface_name(surf_idx);
+		Dictionary lods = Dictionary();
+		Ref<Material> mat = p_mesh->get_surface_material(surf_idx);
+
+		Vector<Vector3> vertex_array = arr[ArrayMesh::ARRAY_VERTEX];
+
+		const int vert_arr_len = vertex_array.size();
+		int i = 0;
+		while (i < vert_arr_len) {
```
Comment (reduz) [inline editor/import/resource_importer_scene.cpp:412]: Why not using for loops?

**Review by reduz — DISMISSED** [review]

**Comment by reduz** [comment]
Seems good for the most part, should be able to merge it

**Comment by fire** [comment]
Will revise the for loop.

**Comment by fire** [comment]
1. scaled the blends shape test by 10 and then apply
2. check scaling on root and the animations play 
3. done

[MorphStressTest.zip](https://github.com/godotengine/godot/files/9491297/MorphStressTest.zip)

image

**Comment by fire** [comment]
> Consider other objects types for rescaling? For example, light attenuation? camera scale/near/far?

Cameras and punctual lights are like points, so scaling a point doesn't do much. I don't know if scaling a light's light-units by distance-scale has good formula.

**Comment by Calinou** [comment]
> Cameras and punctual lights are like points, so scaling a point doesn't do much. I don't know if scaling a light's light-units by distance-scale has a good formula since the physical light units is not default.

Scaling a positional light by increasing its range should work well enough in some cases.

**Comment by fire** [comment]
@reduz can you remove the merge block? The for loop has been removed.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
